### PR TITLE
Fixes for building from source on Win10 with Visual Studio 2022 (17.9.6) for C++20 standard.

### PR DESCRIPTION
**Pull request by akimzav**
Fixes two types of building errors in the VS, and lazily-patches one in .cmake file.
The errors in VS are the following:

1. Multiple instances of "error C2445: result type of conditional expression is ambiguous: types '`const char [..]`' and '`G4Sting`' can be converted to multiple common types". This is fixed by adding explicit construction of G4String from the string literals.
2. One "error LNK2005: "public: virtual void __cdecl TG4MoleculeShoot<class G4Track>::Shoot(class G4MoleculeGun *)" (?Shoot@?$TG4MoleculeShoot@VG4Track@@@@UEAAXPEAVG4MoleculeGun@@@z) already defined in G4MoleculeGun.obj".
This error is due to the fact that the specialization of `TG4MoleculeShoot<TYPE>::Shoot` for `TYPE=G4Track` first appears at `G4MoleculeGun.cc`, but before that, the `TG4MoleculeShoot<TYPE>` class is instantiated for `TYPE=G4Track` as a result of calling the `CreateNewType<G4Track>(shootName)` in file `G4MoleculeGunMessenger.cc (line 84)`; the definition of this templated function contains the `TG4MoleculeShoot<TYPE>`, and the file it's in is a header, and, of course, that file `#include`s the `G4MoleculeGun.hh`.
As such, the `TG4MoleculeShoot<G4Track>` is instantiated BEFORE the specialization `TG4MoleculeShoot<G4Track>::Shoot` appears in `G4MoleculeGun.cc`, and at that point, the specialization is seen as another (multiple) definition to the already instantiated template.
The fix is simply declaring the specialization beforehand, right below the `TG4MoleculeShoot` template. 

The lazy patch relates to the CMake error. The regex in cmake/Modules/G4DeveloperAPI.cmake (line 936) is wrong. CMake 3.29.2 can't comprehend it, and neither can I, so I commented the line out, as it seems like nothing important.

As a result, the Geant4 solution can be build.

### CODE REVIEW

File: cmake/Modules/G4DeveloperAPI.cmake [hunk 1]
```diff
@@ -933,7 +933,7 @@ function(geant4_compose_targets)
     geant4_get_module_property(__parent_target ${__module} PARENT_TARGET)
     get_filename_component(__listdir "${__listfile}" DIRECTORY)
     # Remove generated headers
-    list(TRANSFORM __publichdrs REPLACE "^${PROJECT_BINARY_DIR}/.*$" "")
+    #list(TRANSFORM __publichdrs REPLACE "^${PROJECT_BINARY_DIR}/.*$" "")
```
Comment (drbenmorgan) [inline cmake/Modules/G4DeveloperAPI.cmake:936]: What's the value of `PROJECT_BINARY_DIR` here? The regex is very simple in that is should just remove the value of that variable at the beginning of the input string.

**Comment by drbenmorgan** [comment]
Also, when you configure the project, what CMake arguments are you setting, and what values are printed in the initial info above the compiler when running cmake, e.g.

```
-- The C compiler identification is MSVC 19.36.32532.0
-- The CXX compiler identification is MSVC 19.36.32532.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.36.32532/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.36.32532/bin/Hostx64/x64/cl.exe - skipped
```

File: cmake/Modules/G4DeveloperAPI.cmake (hunk 1 above)
Comment (akimzav) [reply cmake/Modules/G4DeveloperAPI.cmake:936]: The value of PROJECT_BINARY_DIR is D:/Documents/_C++_projects/geant4-v11.2.1-build.
CMake fails on this line with a:
```
[INFO] RegularExpression::compile(): Nested *?+.
[INFO] RegularExpression::compile(): Error in compile.
```

**Comment by drbenmorgan** [comment]
Thanks for the info! We'll set up a C++20 CI build to check that we can reproduce this and for regression testing (though note that we tend to stay a little behind the latest/greatest MSVC version).

On the regex error, it's due to the `++` in the path name - that's not wrong, but is confusing the regex. There's a solution here: https://stackoverflow.com/questions/13569660/cmake-escaping-symbols-inside-a-variable-in-regular-expressions if you want to try adding that, otherwise we can fix upstream separately.

**Comment by akimzav** [comment]
> Thanks for the info! We'll set up a C++20 CI build to check that we can reproduce this and for regression testing (though note that we tend to stay a little behind the latest/greatest MSVC version).
> 
> On the regex error, it's due to the `++` in the path name - that's not wrong, but is confusing the regex. There's a solution here: https://stackoverflow.com/questions/13569660/cmake-escaping-symbols-inside-a-variable-in-regular-expressions if you want to try adding that, otherwise we can fix upstream separately.

Done a dirty but technically allowed fix on that issue.

**Comment by drbenmorgan** [comment]
Patches applied upstream, will be in next release/patch, thank you for the contribution!

**Pull request closed without merging**